File: 수식_최대화.js
function solution(expression) {
    let answer = 0;

    const orders = [
        ['+', '-', '*'],
        ['+', '*', '-'],
        ['-', '+', '*'],
        ['-', '*', '+'],
        ['*', '-', '+'],
        ['*', '+', '-'],
    ];

    let arr = exp2arr(expression);

    for (let order of orders) {
        let newArr = arr.slice();
        order.forEach(operator => {
            newArr = operate(newArr, operator);
        })
        answer = Math.max(answer, Math.abs(newArr[0]))
    }

    console.log(answer);
    return answer;

    function exp2arr(expression) {
        let result = [];
        let temp = '';
        for (let x of expression) {
            if (isNaN(x)) {
                result.push(Number(temp));
                temp = '';
                result.push(x);
            } else {
                temp += x;
            }
        }
        result.push(Number(temp))
        return result;
    }

    function calculate(n1, n2, operator) {
        if (operator === '+') return n1 + n2;
        if (operator === '-') return n1 - n2;
        if (operator === '*') return n1 * n2;
    }

    function operate(arr, operator) {
        const stack = [];
        for (let x of arr) {
            if (stack[stack.length - 1] === operator) {
                stack.pop();
                stack.push(calculate(stack.pop(), x, operator))
            } else {
                stack.push(x);
            }
        }
        return stack;
    }
}


const expression = "100-200*300-500+20"
solution(expression)

File: python/kakao대비/수식_최대화.py
import re

def solution(expression):
    numbers = list(map(int, re.compile(r'[\+\-\*]').split(expression)))
    operators = [x for x in expression if not x.isdecimal()]

    def calculate_with_oper(num1, num2, operator):
        if operator == '+':
            return num1 + num2
        if operator == '-':
            return num1 - num2
        if operator == '*':
            return num1 * num2

    def calculate_with_order(order, numbers, operators):
        for i in range(3):
            prior_operator = order[i]
            new_numbers = [numbers.pop(0)]
            new_operators = []
            while numbers and operators:
                number = numbers.pop(0)
                operator = operators.pop(0)
                if operator == prior_operator:
                    new_numbers[-1] = calculate_with_oper(new_numbers[-1], number, operator)
                else:
                    new_numbers.append(number)
                    new_operators.append(operator)
            numbers = new_numbers
            operators = new_operators
        return abs(numbers[0])

    orders = [
        ['+', '-', '*'],
        ['+', '*', '-'],
        ['*', '+', '-'],
        ['-', '+', '*'],
        ['-', '*', '+'],
        ['*', '-', '+'],
    ]

    answer = 0
    for order in orders:
        answer = max(answer, calculate_with_order(order, [x for x in numbers], [x for x in operators]))
    return answer



solution("100-200*300-500+20")